File: simulator/actions/ranged_attack.py
import math
from ..actions.actoid import FactoryFlags
from ..actions.attack import AttackFactory, Attack
from ..battle_map import Map, map_position_toggled_cache
from ..misc import Visibility
from ..conditions import Conditions, is_affected_by_any, get_swallower
from ..resources import Uses, ResourceRefreshType
from ..utils.roll_types import RollType, ROLL_TYPE_DELTA
import logging
import numba_functions as nf

logger = logging.getLogger("Encounterra")


class RangedAttackFactory(AttackFactory):

    def __init__(self, name, combatant, to_hit, dmg_dice, dmg_bonus, dmg_type, attack_range, action_type, crit_range=1, ammo=Uses(math.inf, ResourceRefreshType.NEVER), on_hit=None, extra_dmg=None, uses_dex=True, to_hit_bonus_die=None, **kwargs):
        super().__init__(name, combatant, to_hit, dmg_dice, dmg_bonus, dmg_type, attack_range, action_type, crit_range, ammo, on_hit, extra_dmg, uses_dex, False, to_hit_bonus_die)
        self.flags |= FactoryFlags.IS_RANGED

    def get_ability_name(self):
        return "Ranged Attack"

    def create(self, target):
        return RangedAttack(target, self)

    def create_all(self, previous_action_in_dag=None):
        targets = self.get_eligible_targets()
        return [RangedAttack(t, self) for t in targets]

    def calculate_threat_to_target(self, target, **kwargs):
        consider_dist = kwargs.get("consider_dist", False)
        roll_type = kwargs.get("roll_type", RollType.STRAIGHT)

        to_hit_total = self.to_hit
        to_hit_total += ROLL_TYPE_DELTA[roll_type][max(0, min(target.ac - to_hit_total, 20))]

        # TODO: Should I include roll types here? There may be a use-case in the future
        battle_map = Map.get()
        if not consider_dist or battle_map.get_cartesian_distance_combatants(self.combatant, target) <= self.range:
            acc = nf.mean_dmg(to_hit_total, self.dmg_dice, self.dmg_bonus, target.ac, target.is_immune_to(self.dmg_type),
                           target.is_resistant_to(self.dmg_type), self.crit_range)
            for extra in self.extra_dmg:
                acc += nf.mean_dmg(to_hit_total, [extra[0]], 0, target.ac,
                                target.is_immune_to(extra[1]), target.is_resistant_to(extra[1]), self.crit_range)
            for oh in self.on_hit:
                acc += nf.calc_p_hit(to_hit_total, target.ac) * oh.calculate_threat(self.combatant, target)
            return acc
        return 0


class RangedAttack(Attack):

    @map_position_toggled_cache
    def calculate_threat(self, **kwargs):
        battle_map = Map.get()
        roll_type = RollType.STRAIGHT if not battle_map.is_enemy_adjacent(self.factory.combatant) else RollType.DISADVANTAGE
        roll_type = RollType.DISADVANTAGE if battle_map.get_cartesian_distance_combatants(self.factory.combatant, self.target) > self.factory.short_range else roll_type
        return self.factory.calculate_threat_to_target(self.target, roll_type=roll_type, **kwargs)

    def clear_cache(self):
        self.calculate_threat.cache_clear()
        self.calculate_threat_delta.cache_clear()
        #self.get_eligible_coords.cache_clear()

    #@map_toggled_cache_with_key(key=lambda self, distances, shortest_paths: hashkey(self.factory.name, tuple(Map.get().get_combatant_position(self.factory.combatant).get()[0])))
    def get_eligible_coords(self, distances, shortest_paths):
        battle_map = Map.get()
        swallower = get_swallower(self.factory.combatant)
        if swallower:
            if swallower is self.target:
                return [tuple(battle_map.get_combatant_position(self.factory.combatant).get()[0])]  # Makes barely any sense but ok
            return None
        curr_coord = tuple(battle_map.get_combatant_position(self.factory.combatant).get()[0])
        # if self.factory.combatant.movement > 0 and not is_affected_by_any(self.factory.combatant, Conditions.GRAPPLED, Conditions.GRAPPLING, Conditions.RESTRAINED):
        if not is_affected_by_any(self.factory.combatant, Conditions.GRAPPLED, Conditions.GRAPPLING, Conditions.RESTRAINED):
            free_coords_in_range = nf.get_free_coords_in_cartesian_range(
                battle_map.grid,
                battle_map.get_combatant_position(self.target).get(),
                distances,
                self.factory.combatant.size.value,
                self.factory.range, self.factory.combatant.id)
            if not battle_map.effect_tracker.is_combatant_hidden_from(self.factory.combatant, self.target):
                return [coord for coord in free_coords_in_range if battle_map.visibility_dict_for_all_coords[coord][self.target] is not Visibility.NONE]
            else:
                # We only consider the coords where Visibility.NONE transitions into any other kind
                ret = list()
                for coord in free_coords_in_range:
                    if battle_map.visibility_dict_for_all_coords[coord][self.target] is not Visibility.NONE:
                        try:
                            if battle_map.visibility_dict_for_all_coords[tuple(shortest_paths[coord])][self.target] is Visibility.NONE:
                                ret.append(coord)
                        except KeyError:
                            ret.append(coord)
                return ret
        elif battle_map.get_cartesian_distance_combatants(self.factory.combatant, self.target) <= self.factory.range and \
                battle_map.visibility_dict_for_all_coords[curr_coord][self.target] is not Visibility.NONE:
            return [curr_coord]
        return None

File: c++/actions/ranged_attack.cpp
#include "actions/ranged_attack.hpp"
#include "core/interfaces.hpp"
#include "core/battle_map.hpp"
#include "core/combatant.hpp"
#include "effects/effect_tracker.hpp"

namespace enc
{
  RangedAttackFactory::RangedAttackFactory(const std::string &name, const std::string &abilityName, Combatant& combatant, AbilityType abilityType,
                                           int toHit, std::vector<Die> dmgDice, int dmgBonus, DamageType dmgType, int attackRange, int critRange,
                                           Uses &&ammo, std::vector<std::unique_ptr<OnHit>> onHit, std::vector<DmgDieWithType> extraDmg, bool usesDex,
                                           bool twoHanded, Die toHitBonusDie)
      : AttackFactory(name, abilityName, combatant, abilityType, toHit, dmgDice, dmgBonus, dmgType, attackRange, critRange, std::move(ammo),
                      std::move(onHit), extraDmg, usesDex, twoHanded, toHitBonusDie)
  {
    setFlag(FactoryFlags::IS_RANGED);
  }

  std::vector<std::shared_ptr<Actoid>> RangedAttackFactory::createAll(void *previousActionInDag)
  {
    auto eligibleTargets = getEligibleTargets();
    std::vector<std::shared_ptr<Actoid>> result;
    result.reserve(eligibleTargets.size());
    for(const auto &target : eligibleTargets)
      {
        result.push_back(std::make_shared<RangedAttack>(AbilityType::RANGED_ATTACK, *target.lock(), *this));
      }
    return result;
  }

  std::shared_ptr<Actoid> RangedAttackFactory::create(void *target)
  {
    return std::make_shared<RangedAttack>(AbilityType::RANGED_ATTACK, *static_cast<Combatant *>(target), *this);
  }

  std::optional<CoordVector>
  RangedAttack::getEligibleCoords(const blaze::DynamicVector<int> &distances, const blaze::DynamicMatrix<Coord> &shortestPaths)
  {
    RangedAttackFactory &factory = dynamic_cast<RangedAttackFactory &>(getFactory());
    BattleMap &battleMap = BattleMap::getInstance();
    auto combatant = factory._combatant;
    Coord currCoord = battleMap.getCombatantCoordinates(*combatant).getRoot();

    if(Combatant *swallower = combatant->getSwallower())
      {
        if(*swallower == _target)
          {
            return CoordVector{currCoord};
          }
        return {};
      }

    if(!combatant->isAffectedByAny({Conditions::GRAPPLED, Conditions::GRAPPLING, Conditions::RESTRAINED}))
      {
        CoordVector freeCoordsInRange
          = battleMap.getFreeCoordsInCartesianRange(battleMap.getCombatantCoordinates(_target).get(), distances, combatant->getSize(),
                                                    factory._attackRange, combatant->_instanceId);

        if(!EffectTracker::getInstance().isCombatantHiddenFrom(*combatant, _target))
          {
            CoordVector visibleCoords;
            std::copy_if(freeCoordsInRange.begin(), freeCoordsInRange.end(), std::back_inserter(visibleCoords),
                         [&battleMap, this](const Coord &coord) {
                           return battleMap.getVisibilityFromCoord(coord, _target) != Visibility::NONE;
                         });
            return visibleCoords;
          }
        else
          {
            // We only consider the coords where Visibility::NONE transitions into any other kind
            CoordVector transitionCoords;
            for(const auto &coord : freeCoordsInRange)
              {
                if(battleMap.getVisibilityFromCoord(coord, _target) != Visibility::NONE)
                  {
                    try
                      {
                        if(battleMap.getVisibilityFromCoord(shortestPaths(coord[0], coord[1]), _target) == Visibility::NONE)
                          {
                            transitionCoords.push_back(coord);
                          }
                      }
                    catch(const std::out_of_range &)
                      {
                        transitionCoords.push_back(coord);
                      }
                  }
              }
            return transitionCoords;
          }
      }
    else if(battleMap.getCartesianDistanceCombatants(*combatant, _target) <= factory._attackRange
            && battleMap.getVisibilityFromCoord(currCoord, _target) != Visibility::NONE)
      {
        return CoordVector{currCoord};
      }

    return {};
  }

  size_t RangedAttack::hash() const
  {
    size_t h = std::hash<int>{}(static_cast<int>(getAbilityType()));
    h ^= std::hash<int>{}(static_cast<int>(getFlags())) + 0x9e3779b9 + (h << 6) + (h >> 2);
    h ^= std::hash<int>{}(_target._instanceId) + 0x9e3779b9 + (h << 6) + (h >> 2);
    h ^= std::hash<std::string>{}(_factory._name) + 0x9e3779b9 + (h << 6) + (h >> 2);
    return h;
  }

  bool RangedAttack::equals(const Actoid &other) const
  {
    if(auto *rangedAttack = dynamic_cast<const RangedAttack *>(&other))
      {
        return getAbilityType() == other.getAbilityType() && getFlags() == other.getFlags()
               && _target._instanceId == rangedAttack->_target._instanceId && _factory._name == rangedAttack->_factory._name;
      }
    return false;
  }
}